stablehlo/dialect/TypeInference.cpp

```diff
@@ -2670,10 +2670,27 @@ LogicalResult inferSelectOp(
                                "requires the same shape for all operands");
 
   // The output shape should be derived from the most specific parts of the
-  // `onTrue` and `onFalse` (see documentation for details).
-  SmallVector<Type> inferredReturnTypes;
-  return inferMostSpecificTypeComponents(location, {trueType, falseType},
-                                         inferredReturnShapes);
+  // `onTrue` and `onFalse` when the bounds are missing (see documantation for
+  // `inferMostSpecificTypeComponents` for details).
+  auto onTrueRankedTy = onTrue.getType().dyn_cast<RankedTensorType>();
+  auto onFalseRankedTy = onFalse.getType().dyn_cast<RankedTensorType>();
+  if (!onTrueRankedTy || !onFalseRankedTy ||
+      (encodingToBounds(onTrueRankedTy.getEncoding()).empty() &&
+       encodingToBounds(onFalseRankedTy.getEncoding()).empty()))
+    return inferMostSpecificTypeComponents(location, {trueType, falseType},
+                                           inferredReturnShapes);
```

Comment:
Hmm I'm not sure I follow. If we want to infer the least specific type, why do we need this code block that infers the most specific type under some circumstances?

Reply:
For this part of the check, since previous code was inferring the most specific shape before the bounds were introduced, I followed suit to infer the most specific shape when there are no static bounds. I intended to infer the least specific type only when the bounds are present. We end up having most specific shape w/o bounds and least specific shape/bounds if bounds are present. Does this pose issues down the line?

Reply:
"Does this pose issues down the line?". Since we don't have a formal calculus for type inference yet (theoretical part), and since type inference with dynamic shapes is much less used than type inference with static shapes (practical part), it's hard to answer this definitively.

Based on my experience with working on programming languages and type systems, I would find this behavior quite surprising. Intuitively, we'd want to treat unbounded and bounded dynamism as similarly as possible, and this code would contradict this intuition. Also, when bounds are present, the behavior seems to flip for all dims (not just for bounded dims, but for unbounded dims), which introduces an additional element of surprise.

Overall, I would recommend that we stick to unsurprising solutions / solutions that align with other programming languages, unless there's a very good reason to deviate.